### ISSUE TITLE
Avoid name collisions between object methods and schema/datapackage attributes

### ISSUE THREAD

**Issue by vitorbaptista** [okfn/datapackage-py#3]
Consider the following example:

```python
import datapackage

dp = datapackage.DataPackage({
    'schema': '...'
})
dp.schema  # What is it?
```

Reading the code, I would expect `dp.schema == '...'`, but `dp.schema` is actually a `datapackage.schema.Schema` instance. This happens because the methods defined in the `DataPackage` class overloads whatever it is in the `datapackage.json`. The only ways I can see to fix it is to either:

1. Have no methods defined in `DataPackage`;
2. Don't allow accessing attributes as properties (i.e. `dp.schema`).

As `1` is very restritive, I'm leaning towards `2`. We could implement `2` as `sqlalchemy` used to do, having a magical `.c` property that allows accessing as `dp.c.schema`, or defining getter/setter methods like `dp.get('schema')` and `dp.set('schema', 'foo')`, or even exposing a dict (or dict-like) interface, as `dp['schema'] = 'foo'`.

Even without this issue, we would have to have another method of accessing the methods, as the datapackage's attributes don't need to be valid method names. They could be `my-attribute`, or `$schema`, or `my attribute`, or have unicode characters like `órgão` (public body), etc..

Given those issues, I think the best solution is using `dp.get()` and `dp.set()` methods.

**Comment by femtotrader** [comment]
getter / setter are not Pythonic

http://stackoverflow.com/questions/2627002/whats-the-pythonic-way-to-use-getters-and-setters

**Comment by pwalsh** [comment]
Getters and setters are not Pythonic, but practicality beats purity :).

In light of @vitorbaptista's observations, the nature of the problem here requires a solution that namespaces the data, and exposes it as a dict rather than methods.

[WTForms](https://wtforms.readthedocs.org/en/latest/forms.html#wtforms.form.Form.data) deals with a similar problem, and exposes the data object on a `.data` property, which is a dict.

In summary, I'd suggest we follow that pattern.

```
import datapackage

dp = datapackage.DataPackage({
    'schema': '...'
})
# data is exposes on dp.data
dp.data.schema
# methods that operate on the data are exposed as normal
dp.as_json()
dp.validate()
```

**Comment by femtotrader** [comment]
What is your opinion about overloading `__setitem__` and `__getitem__` ?

Instead of dict it could be useful to use dot accessible dict like bunch / lunch / munch

https://github.com/Infinidat/munch

**Comment by pwalsh** [comment]
@femtotrader not sure that works with unicode keys on the dict, and doesn't solve the core problem that other method names might override the names on the datapackage itself.

**Comment by femtotrader** [comment]
You are right about the risk that other method names override the names on the datapackage itself.

But munch is unicode ready

```
In [1]: import munch
In [2]: b=munch.Munch()
In [3]: b['\ua000\ub000\uc000']='\ud000\ua000\ub000'
In [4]: b
Out [5]: Munch(ꀀ뀀쀀='퀀ꀀ뀀')
In [6]: b.ꀀ뀀쀀
Out[7]: '퀀ꀀ뀀'
In [8]: b['ꀀ뀀쀀']
Out[8]: '퀀ꀀ뀀'
```

**Comment by vitorbaptista** [comment]
It might be me getting older, but I'm starting to avoid magic :smile:

**Comment by vitorbaptista** [comment]
Oops! Closed this by mistake.

**Comment by vitorbaptista** [comment]
I think it's too hard to use dot accessible attributes while keeping consistency. For example, if I'm able to do `datapackage.license` I would expect to also be able to do `datapackage.license.type` as well. To implement that, we would have to use Munch not only for the top-level attributes, but also for every nested dict.

Also, I find a bit surprising that `datapackage.license = { 'type': 'foo', 'url': 'bar' }` would change the dict I passed. Overloaded operators is cool, but it can create many hard-to-track bugs. For example:

```python
datapackage.license = { 'type': 'foo', 'url': 'bar' }
assert type(datapackage.license) == dict  # BOOM!
```

This isn't a very good example (as the user should use `isinstance` instead) but still, "Explicit is better than implicit."

I don't think the gain from being able to write `datapackage.license` instead of `datapackage['license']` is work it.

**Comment by femtotrader** [comment]
Munch (Bunch / Lunch) have a method called munchify

```
In [34]: import munch

In [35]: d = {
   ....:     'license': {
   ....:         'type': 'foo',
   ....:         'url': 'bar'
   ....:     }
   ....: }
In [36]: data = munch.munchify(d)

In [37]: data
Out[37]: Munch(license=Munch(type='foo', url='bar'))

In [38]: data.license
Out[38]: Munch(type='foo', url='bar')

In [39]: data.license.type
Out[39]: 'foo'

In [40]: data['license']['url']
Out[40]: 'bar'
```

**Comment by pwalsh** [comment]
@femtotrader I understand the general implementation. I haven't used mulch but have used this pattern in other libs. I'm just not sure we really want / need such a dependency here. 

@vitorbaptista what about just exposing data as a dict on .data as I show above? Consumers can just get and set there like any python dict, and if you like, for convenience you could expose explicit get and set methods on the instance as you demonstrate.

**Comment by vitorbaptista** [comment]
@pwalsh I'm a bit concerned with exposing the internal data dict directly to the users, as this lack of encapsulation will make things difficult to change afterwards. Still, that's the best option in my opinion. I'll implement it :+1:

**Comment by pwalsh** [comment]
:+1:
